include/odp/api/spec/packet.h

```diff
@@ -2140,6 +2141,66 @@ uint32_t odp_packet_payload_offset(odp_packet_t pkt);
  */
 int odp_packet_payload_offset_set(odp_packet_t pkt, uint32_t offset);
 
+/** ODP packet proto stats op */
+typedef enum odp_packet_proto_stats_op_t {
+	/** Set inner octet adjust to total packet len
+	 * Inner octet count for proto stats is computed as
+	 * `odp_packet_len() + INNER_ADJ`
+	 */
+	ODP_PACKET_PROTO_STATS_OP_S16_INNER_ADJ,
+	/** Set outer octet adjust to total packet len
+	 * Outer octet count for proto stats is computed as
+	 * `odp_packet_len() + OUTER_ADJ`
+	 */
+	ODP_PACKET_PROTO_STATS_OP_S16_OUTER_ADJ,
+	/** Enable / Disable Outer octet update for a packet */
+	ODP_PACKET_PROTO_STATS_OP_BOOL_OUTER_ENABLE,
+} odp_packet_proto_stats_op_t;
+
+/**
+ * Set proto stats object.
+ *
+ * Associate a proto stats object to a packet. This enables updating protocol
+ * stats of the packet in the given object after packet is enqueued to queue
+ * and reaches a terminal state i.e either transmit or drop.
+ *
+ * @param pkt   Packet handle
+ * @param stat  Proto stats object handle
+ */
+void odp_packet_proto_stats_set(odp_packet_t pkt, odp_proto_stats_t stat);
+
+/**
+ * Get proto stats object.
+ *
+ * Get the proto stats object associated with the given packet.
+ *
+ * @param pkt Packet handle
+ *
+ * @return Proto stats object handle or ODP_PROTO_STATS_INVALID if not set.
+ */
+odp_proto_stats_t odp_packet_proto_stats(odp_packet_t pkt);
+
+/**
+ * Set packet specific metadata for proto stats
+ * Per-Packet metadata needs to be updated for modes ODP_PROTO_STATS_MODE_TX_*
+ *
+ * @param pkt   Packet handle
+ * @param op    Proto stats meta data op
+ * @param data  Pointer to meta data specific to op
+ *
+ * @see odp_proto_stats_mode_t
+ */
+void odp_packet_proto_stats_ops_set(odp_packet_t pkt, odp_packet_proto_stats_op_t op, void *data);
```

Comment:
How is one supposed to use the data pointer? The object pointed by it does not seem to be defined anywhere.

Reply:
I missed to document this more in enum's of ``odp_packet_proto_stats_op_t ``
When ``op`` is ODP_PACKET_PROTO_STATS_OP_**S16**_INNER_ADJ data is supposed to be pointer to **S16** i.e int16_t.
When ``op`` is ODP_PACKET_PROTO_STATS_OP_**BOOL**_OUTER_ENABLE data is supposed to be pointer to bool.

Reply:
I am not sure this is really in line with the rest of ODP API that does not use void pointers like that but instead uses specified types (c.f. odp_ipsec_test_sa_param_t, odp_pmr_param_t, odp_tm_egress_t, odp_threshold_t).

Reply:
Ack, will change to individual calls to set the content.